crypto-sodium/lib/Crypto/Sodium/Salt.hs

```diff
@@ -0,0 +1,73 @@
+-- SPDX-FileCopyrightText: 2021 Serokell
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | This module gives different ways of obtaining salts.
+--
+-- A “salt” is additional input provided to certain functions.
+-- Unlike a nonce, salt is used for “namespacing”, so there is not
+-- requirement that it is used only once, instead you use different
+-- salts to deisgnate different applications of an algorithm.
+--
+-- This is a subtle and pure semantical difference, so, for convenience,
+-- we also re-export some functions from "Crypto.Sodium.Nonce".
+module Crypto.Sodium.Salt
+  (
+  -- * Literals
+    utf8
+
+  -- * Random salt generation
+  , Crypto.Sodium.Nonce.generate
+  ) where
+
+import Data.ByteArray.Sized (SizedByteArray, unsafeSizedByteArray)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Internal as BS
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import Language.Haskell.TH (Exp, Q)
+import qualified Language.Haskell.TH.Lib as TH
+import Language.Haskell.TH.Quote (QuasiQuoter (..))
+import qualified Language.Haskell.TH.Syntax as TH
+
+import qualified Crypto.Sodium.Nonce
+
+
+-- | Make a /sized/ 'ByteString' from a type-level string literal.
+--
+-- @
+-- {-# LANGUAGE QuasiQuotes #-}
+--
+-- {- ... -}
+--
+-- let salt = [utf8|hello world|] :: 'SizedByteArray' 11 ByteString
+-- @
+--
+-- This uses Template Haskell to compute the length of the UTF-8 encoding
+-- of the string you provide. Note that the string will be taken as is,
+-- with all whitespace preserved, for example @[utf8| x |]@ will have length 3.
+utf8 :: QuasiQuoter
+utf8 = QuasiQuoter { quoteExp, quotePat, quoteType, quoteDec }
+  where
+    quoteExp :: String -> Q Exp
+    quoteExp str =
+        let
+          valExpr = [e|BS.packBytes $(TH.lift $ BS.unpackBytes bs)|]
```

Comment:
We can optimize this quite a bit, if you don't mind unsafe primitives.
```suggestion
          valExpr = [|unsafePerformIO (BS.unsafePackAddressLen len $cstr)|]
          cstr = TH.litE . TH.stringPrimL $ BS.unpack bs
          -- this constructs an Addr# literal from a [Word8]
```
(where `unsafePackAddressLen` is from `Data.ByteString.Unsafe`; this code should work since GHC 7.6 IIRC, before that `stringPrimL` accepted `[Char]`; note that `lift` is applied to `len` automatically, you almost never need to use it explicitly)

This code will be constant-time in the resulting compiled binary, while the one using `packBytes` is O(len). Also, the constant factor is actually quite small here, since it just wraps a raw address in a `ByteString`.

... or you could just depend on `th-lift-instances`, which defines a `Lift` instance for `ByteStrings` along these lines.

Reply:
Ah yes, well, actually, I was going to use [`unsafePackLiteral`](https://hackage.haskell.org/package/bytestring-0.11.1.0/docs/Data-ByteString-Internal.html#v:unsafePackLiteral), but it turned out that it is only available in the very latest version of `bytestring`, so I reverted to the dumbest solution.

I guess I’ll try something in the middle then.